packages/http-client/src/middleware/middleware-type.ts

```diff
@@ -0,0 +1,66 @@
+import type { Destination } from '@sap-cloud-sdk/connectivity';
+import { createLogger } from '@sap-cloud-sdk/util';
+
+const logger = createLogger('middleware');
+
+/**
+ * In/out parameter in the chain of middlewares.
+ */
+export interface MiddlewareInOut<T> {
+  /**
+   * Function execute inside the middleware.
+   */
+  fn: () => Promise<T>;
+  /**
+   * Context of the execution.
+   */
+  context: Context;
+}
+
+/**
+ * Middleware type.
+ */
+export type Middleware<T> = (options: MiddlewareInOut<T>) => MiddlewareInOut<T>;
+
+/**
+ * Context of the middleware.
+ */
+export interface Context {
+  /**
+   * Request category.
+   */
+  category: 'xsuaa' | 'destination' | 'user-defined';
+  /**
+   * Destination used in the request.
+   */
+  destination: Destination;
+  /**
+   * JWT used in the request.
+   */
+  jwt?: string;
+  /**
+   * Arguments used in the request.
+   */
+  args: any[];
+}
+
+/**
+ * Helper function to join a list of middlewares given an initial input.
+ * @param middleWares - Middlewares to be layered around the function.
+ * @param initial - Input for the layering process.
+ * @returns Function with middles wares layered around it.
+ * @internal
+ */
+export function wrapFunctionWithMiddleware<T>(
+  middleWares: Middleware<T>[] | undefined,
+  initial: MiddlewareInOut<T>
+): () => Promise<T> {
+  if (!middleWares || middleWares.length === 0) {
+    return initial.fn;
```

Comment:
```suggestion
    return initial.fn();
```
[q] Since `fn` contains a wrapper function (at least usually), it should probably remove this when no middleware is added, right?

Example before middleware:
```
return execute({});
```
Example after middleware:
```
return wrapFnWithMiddleware(undefined, { fn: () => execute({}) });
```
is currently equal to 
```
return () => execute({});
```

Reply:
Is there any reason why this function should return a function and not call the function directly? Seems like a user would always call it like this: `wrapFnWithMiddleware(foo, bar)();`

Reply:
I guess this may be used for stacking multiple middlewares, but since this function is currently only used in one place it is unclear to me...

Reply:
Yes this could execute the function right away.